[WPE][Debug] ASSERTION FAILED: inProgrammaticScroll == (options.type == ScrollType::Programmatic)

obyknovenius opened the pull request:
#### 50aa99222474489e8be5ff75ffeef6ba20a1078b
<pre>
[WPE][Debug] ASSERTION FAILED: inProgrammaticScroll == (options.type == ScrollType::Programmatic)
<a href="https://bugs.webkit.org/show_bug.cgi?id=259912">https://bugs.webkit.org/show_bug.cgi?id=259912</a>

Reviewed by Simon Fraser.

In `AsyncScrollingCoordinator::requestScrollToPosition()` we assert
that the `scrollableArea`&apos;s `currentScrollType()` matches the
`ScrollPositionChangeOptions`&apos;s type.

In some circumstances, this method can be called as a result of
calling `ScrollableArea::scrollToPositionWithAnimation()`, which
currently doesn&apos;t set its `currentScrollType` from `options` argument.

`LocalFrameView`, which is a subclass of `ScrollableArea`, adds a method
whose only job is to set/restore `currentScrollType` and call parent&apos;s
method. We can move this logic to the `ScrollableArea` class. There&apos;s
actually a FIXME about this.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToPositionWithAnimation): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollToPositionWithAnimation):

Canonical link: <a href="https://commits.webkit.org/270820@main">https://commits.webkit.org/270820@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/00556e7b57088564c71119f12461603710c7a355

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26149 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4759 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23938 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26474 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2172 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23975 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22501 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3210 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23464 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28820 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23852 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29514 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3259 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27406 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4667 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6363 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3726 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3581 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -176,12 +176,23 @@ void ScrollableArea::scrollToPositionWithAnimation(const FloatPoint& position, c
 {
     LOG_WITH_STREAM(Scrolling, stream << "ScrollableArea " << this << " scrollToPositionWithAnimation " << position);
 
+    if (position == scrollPosition())
```
I think this early return needs to be after the call to `scrollAnimator().cancelAnimations()`. There might be an enqueued scroll animation that takes us away from the current scroll position.

obyknovenius replied:
Done

obyknovenius committed:
[WPE][Debug] ASSERTION FAILED: inProgrammaticScroll == (options.type == ScrollType::Programmatic)
https://bugs.webkit.org/show_bug.cgi?id=259912

Reviewed by Simon Fraser.

In `AsyncScrollingCoordinator::requestScrollToPosition()` we assert
that the `scrollableArea`'s `currentScrollType()` matches the
`ScrollPositionChangeOptions`'s type.

In some circumstances, this method can be called as a result of
calling `ScrollableArea::scrollToPositionWithAnimation()`, which
currently doesn't set its `currentScrollType` from `options` argument.

`LocalFrameView`, which is a subclass of `ScrollableArea`, adds a method
whose only job is to set/restore `currentScrollType` and call parent's
method. We can move this logic to the `ScrollableArea` class. There's
actually a FIXME about this.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToPositionWithAnimation): Deleted.
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollToPositionWithAnimation):

Canonical link: https://commits.webkit.org/270820@main
